### PR TITLE
EVG-7833 volumes ui

### PR DESCRIPTION
**Pull request by ybrill**
Volumes are added to the spawnhosts page. Volumes can be added or removed from hosts and deleted. Their display names can be changed (similar to spawnhosts). 
The selected resource type (hosts vs. volumes) and the id are added to the url and volumes link to their hosts and vice versa (for home volumes).

Other modifications to volumes, such as extending their expiration, will be added later.

Screen Shot 2020-04-23 at 5 58 04 PM

### CODE REVIEW

File: service/spawn.go [hunk 1]
```diff
@@ -469,3 +467,100 @@ func (uis *UIServer) modifySpawnHost(w http.ResponseWriter, r *http.Request) {
...
+			uis.LoggedError(w, r, http.StatusBadRequest, errors.Wrapf(err, "volume '%s' is not attached", vol.ID))
+			return
+		}
+		h, err := host.FindOneId(vol.Host)
+		if err != nil {
+			uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "can't get host '%s' for volume '%s'", vol.Host, vol.ID))
+			return
+		}
+		if h == nil {
+			uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "host '%s' for volume '%s' doesn't exist", vol.Host, vol.ID))
```
Comment (ablack12) [inline service/spawn.go:551]: If the host doesn't exist, should we remove vol.Host?

**Comment by ablack12** [comment]
Can you add a way to create volumes?

**Comment by ybrill** [comment]
I plan to add it later. I'll make a separate ticket

**Comment by ybrill** [comment]
https://jira.mongodb.org/browse/EVG-7262

File: service/spawn.go [hunk 2]
```diff
@@ -469,3 +467,100 @@ func (uis *UIServer) modifySpawnHost(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	// take the specified action
+	switch *updateParams.Action {
+	case VolumeRename:
+		uis.LoggedError(w, r, http.StatusUnauthorized, errors.Wrapf(vol.SetDisplayName(*updateParams.NewName), "can't set display name of '%s' to '%s'", vol.ID, *updateParams.NewName))
+
+	case VolumeExtendExpiration:
+		//TODO
```
Comment (ablack12) [inline service/spawn.go:517]: Is there a ticket for this?

File: public/static/partials/user_host_details.html [hunk 3]
```diff
@@ -38,12 +38,15 @@ <h2 style="display:inline-block; padding-right:15px">
...
             </div>
-            <div class="entry"ng-show="curHostData.zone">
+            <div class="entry" ng-show="curHostData.zone">
               <strong>Availability Zone</strong> <span>[[curHostData.zone]]</span>
             </div>
             <div class="entry" ng-show="curHostData.is_virtual_workstation && curHostData.status == 'running'">
               <strong>IDE</strong> <span> <a href="/host/[[curHostData.id]]/ide/" target="_blank">Open IDE</a> </span>
             </div>
+            <div class="entry" ng-show="curHostData.home_volume_id">
+              <strong>Home Volume</strong> <span> <a class="pointer" ng-click="goToVolume(curHostData.home_volume_id)"> [[curHostData.home_volume_id]] </a> </span>
```
[the remark this answers is not in the document]
Comment (ybrill) [reply public/static/partials/user_host_details.html]: done

File: public/static/partials/user_host_options.html [hunk 4]
```diff
@@ -99,13 +99,13 @@
...
+              <a role="menuitem" ng-click="setHomeVolume()">
                 New Volume
               </a>
             </li>
-            <li role="presentation" ng-repeat="volume in $parent.availableVolumes">
-              <a role="menuitem" ng-click="setVolume(volume)">
-                [[ volume.display_name ? volume.display_name : volume.volume_id ]]
+            <li role="presentation" ng-repeat="volume in $parent.volumes | filter:{status:'free'}">
+              <a role="menuitem" ng-click="setHomeVolume(volume)">
+                [[ volume.volume_id + (volume.display_name ? " (" + volume.display_name + ")": "" )]]
```
[the remark this answers is not in the document]
Comment (ybrill) [reply public/static/partials/user_host_options.html]: done (in a new function)

File: service/spawn.go (hunk 1 above)
Comment (ybrill) [reply service/spawn.go:551]: done

File: service/spawn.go [hunk 5]
```diff
@@ -469,3 +467,107 @@ func (uis *UIServer) modifySpawnHost(w http.ResponseWriter, r *http.Request) {
...
+		uis.LoggedError(w, r, http.StatusUnauthorized, errors.New("not authorized to modify this volume"))
+		return
+	}
+
+	if updateParams.Action == nil {
+		uis.LoggedError(w, r, http.StatusBadRequest, errors.New("no action specified"))
+		return
+	}
+
+	mgrOpts := cloud.ManagerOpts{
```
Comment (ablack12) [inline service/spawn.go:501]: Can you do this only if the action isn't VolumeRename?

File: service/spawn.go (hunk 2 above)
Comment (ybrill) [reply service/spawn.go:517]: https://jira.mongodb.org/browse/EVG-7845?filter=26459

File: service/spawn.go (hunk 5 above)
Comment (ybrill) [reply service/spawn.go]: done

**Comment by ablack12** [comment]
Sorry for the one off comments: it looks like the display name logic might have a bug.
image

Also is it possible for this to split into two lines if it's too long for the box? I wouldn't know how to do that though haha so fine either way. Also the buttons should be in a row.

**Comment by ybrill** [comment]
fixed the display name logic.
fixed the alignment of the buttons.

**Pull request merged**